From 16f7ae314877749179d5ed89d45575a6f2160bc6 Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Fri, 28 Jun 2013 16:03:13 +0300
Subject: [PATCH 218/228] be/tx: add grp cblk magic, add tx_group_init_state().

* add group commit block magic,
* move common part of tx_group_init() and tx_group_alloc() into
  tx_group_init_state() function
* change tg_state from bool to enum
---
 be/tx_group.c | 42 +++++++++++++++++++++++-------------------
 be/tx_group.h |  5 ++---
 mero/magic.h  |  6 ++++--
 3 files changed, 29 insertions(+), 24 deletions(-)

diff --git a/be/tx_group.c b/be/tx_group.c
index 27eaa70..d16dc52 100644
--- a/be/tx_group.c
+++ b/be/tx_group.c
@@ -44,22 +44,34 @@ M0_TL_DESCR_DEFINE(tg, "groups", M0_INTERNAL, struct m0_be_tx_group,
 
 M0_TL_DEFINE(tg, M0_INTERNAL, struct m0_be_tx_group);
 
+/* Initialize tx group in the given state excluding its STOB i/o structure. */
+static void tx_group_init_state(struct m0_be_tx_group *gr,
+				struct m0_be_tx_engine *eng,
+				enum m0_be_tg_state state)
+{
+	m0_mutex_init(&gr->tg_lock);
+	gr->tg_hdr.gh_id        = 0;
+	gr->tg_hdr.gh_lsn       = 0ULL;
+	gr->tg_hdr.gh_tx_nr     = 0;
+	gr->tg_hdr.gh_tx_size   = 0;
+	gr->tg_hdr.gh_magic     = M0_TRACE_TG_HDR_MAGIC;
+	gr->tg_cblk.gc_id       = 0;
+	gr->tg_cblk.gc_md5.u_hi = 0;
+	gr->tg_cblk.gc_md5.u_lo = 0;
+	gr->tg_cblk.gc_magic    = M0_TRACE_TG_CBLK_MAGIC;
+	gr->tg_eng              = eng;
+	gr->tg_state            = state;
+	m0_forall(i, ARRAY_SIZE(gr->tg_txs),
+		  (tg_tlist_init(&gr->tg_txs[i]), true));
+	tg_tlink_init_at(gr, &eng->te_tgs[gr->tg_state]);
+}
 
 /* Must be called with eng lock hold.*/
 M0_INTERNAL void tx_group_init(struct m0_be_tx_group *gr,
 			       struct m0_be_tx_engine *eng)
 {
-	m0_mutex_init(&gr->tg_lock);
-	gr->tg_hdr.gh_magic   = M0_TRACE_TG_HDR_MAGIC;
-	gr->tg_hdr.gh_tx_nr   = 0;
-	gr->tg_hdr.gh_tx_size = 0;
-	gr->tg_eng            = eng;
-	gr->tg_lsn            = 0ULL;
-	gr->tg_state          = M0_BGS_INIT;
-	m0_forall(i, ARRAY_SIZE(gr->tg_txs),
-		  (tg_tlist_init(&gr->tg_txs[i]), true));
 	m0_stob_io_init(&gr->tg_stobio);
-	tg_tlink_init_at(gr, &eng->te_tgs[gr->tg_state]);
+	tx_group_init_state(gr, eng, M0_BGS_INIT);
 }
 
 M0_INTERNAL struct m0_be_tx_group *tx_group_alloc(struct m0_be_tx_engine *eng)
@@ -89,15 +101,7 @@ M0_INTERNAL struct m0_be_tx_group *tx_group_alloc(struct m0_be_tx_engine *eng)
 	if (iv->iv_index == NULL)
 		goto free_bv;
 
-	m0_mutex_init(&gr->tg_lock);
-	gr->tg_hdr.gh_tx_nr   = 0;
-	gr->tg_hdr.gh_tx_size = 0;
-	gr->tg_eng            = eng;
-	gr->tg_lsn            = 0ULL;
-	gr->tg_state          = M0_BGS_EMPTY;
-	m0_forall(i, ARRAY_SIZE(gr->tg_txs),
-		  (tg_tlist_init(&gr->tg_txs[i]), true));
-	tg_tlink_init_at(gr, &eng->te_tgs[gr->tg_state]);
+	tx_group_init_state(gr, eng, M0_BGS_EMPTY);
 	return gr;
 
 free_bv:
diff --git a/be/tx_group.h b/be/tx_group.h
index 3c48a2e..9e9a9e0 100644
--- a/be/tx_group.h
+++ b/be/tx_group.h
@@ -59,6 +59,7 @@ struct m0_be_tg_hdr {
 };
 /** On-disk group commit block. */
 struct m0_be_tg_cblk {
+	uint64_t                gc_magic;
 	uint64_t                gc_id;    /* id of last transaction in group. */
 	struct m0_uint128       gc_md5;   /* checksum of all tx grp data
 					     excluding m0_be_tg_cblk. */
@@ -114,9 +115,7 @@ struct m0_be_tx_group {
 	uint64_t                tg_magic;
 	/** Linkage in the m0_be_tx_engine::te_tgs[...] lists. */
 	struct m0_tlink         tg_linkage;
-	/** lsn of transaction group header in the log. */
-	m0_bindex_t             tg_lsn;
-	bool                    tg_state;
+	enum m0_be_tg_state     tg_state;
 	/** Per-state lists of transactions in this group. */
 	struct m0_tl            tg_txs[M0_BTS_NR];
 	/** Protects all fields of this struct. */
diff --git a/mero/magic.h b/mero/magic.h
index 7b251d1..59a4511 100644
--- a/mero/magic.h
+++ b/mero/magic.h
@@ -728,14 +728,16 @@ enum m0_magic_satchel {
 	M0_TRACE_TX_LIST_MAGIC       = 0x3311FE1E556E1277,
 	/* m0_be_tx tx group list head magic (codified bee)  */
 	M0_TRACE_TX_GROUP_LIST_MAGIC = 0x33c0d1f1edbee377,
-	/* m0_be_tx_group::tg_magic for tg_linkage (starshine) */
+	/* m0_be_tx_group::tg_magic for tg_linkage (Starshine) */
 	M0_TRACE_TG_MAGIC            = 0x3357a35412e77777,
-	/* m0_be_tx_engine::te_groups head magic (double bass) */
+	/* m0_be_tx_engine::te_groups head magic (Double Dass) */
 	M0_TRACE_TG_HEAD_MAGIC       = 0x33d08b1eba557777,
 	/* m0_be_tx_hdr magic (19-2000) */
 	M0_TRACE_TX_HDR_MAGIC        = 0x3319200077777777,
 	/* m0_be_tg_hdr magic (Clint Eastwood) */
 	M0_TRACE_TG_HDR_MAGIC        = 0x33c1127ea57300d7,
+	/* m0_be_tg_cblk magic (Feel Good Inc.) */
+	M0_TRACE_TG_CBLK_MAGIC       = 0x33fee1600d12c777,
 };
 
 #endif /* __MERO_MERO_MAGIC_H__ */
-- 
1.8.3.2

